### PR TITLE
Vlad rossohin/user profile

### PR DESCRIPTION
**Pull request by VladRossohin**
Please, review my code. It's almost ready. Thanks!

### CODE REVIEW

File: DIMS/DIMS.BL/DTO/TaskDTO.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
+        public TaskDTO()
+        {
+            this.UserTasks = new HashSet<UserTaskDTO>();
+        }
+
+        public int TaskId { get; set; }
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public System.DateTime StartDate { get; set; }
```
Comment (maximsan) [inline DIMS/DIMS.BL/DTO/TaskDTO.cs]: remove System

File: DIMS/DIMS.BL/Interfaces/IUserService.cs [hunk 2]
```diff
@@ -14,5 +14,8 @@ public interface IUserService : IDisposable
         Task<OperationDetails> Create(UserDTO userDto);
         Task<ClaimsIdentity> Authenticate(UserDTO userDto);
         Task SetInitialData(UserDTO adminDto, List<string> roles);
+        void DeleteUserByEmail(string email);
+
+        //Task GetUserById(int id);
```
Comment (maximsan) [inline DIMS/DIMS.BL/Interfaces/IUserService.cs]: delete

File: DIMS/DIMS.BL/Services/UserProfileService.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace HIMS.BL.Services
+{
+    public class UserProfileService : IUserProfileService
+    {
+
+        private IUnitOfWork DimsDatabase { get; }
+        private IProcedureManager Pm { get; }
+        private UserService _userService { get; }
+        private UserProfileRepository _repository { get; }
```
Comment (maximsan) [inline DIMS/DIMS.BL/Services/UserProfileService.cs]: Let's give them name according to convention. If it's property it must be always PascalCase, but if it's field it should be camelCase. If it's private **field** you can start it from _

File: DIMS/DIMS.BL/Services/UserProfileService.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            // Validation (must be improve)
+            if (userProfile.Name.Length > 25)
+                throw new ValidationException($"The length of {nameof(userProfile.Name)} must be less than 25"
+                    , nameof(userProfile.Name));
+            if (userProfile.LastName.Length > 25)
+                throw new ValidationException($"The length of {nameof(userProfile.LastName)} must be less than 25"
+                    , nameof(userProfile.LastName));
+            if (userProfile.Address.Length > 255)
+                throw new ValidationException($"The length of {nameof(userProfile.Address)} must be less than 25"
+                    , nameof(userProfile.Address));
```
Comment (maximsan) [inline DIMS/DIMS.BL/Services/UserProfileService.cs]: Let's do this validation on user profile model in controller using attributes

File: DIMS/DIMS.DAL.Data/UserTask.cs [hunk 5]
```diff
@@ -11,14 +11,26 @@ namespace HIMS.EF.DAL.Data
...
+
     public partial class UserTask
     {
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
+        public UserTask()
+        {
+            this.TaskTracks = new HashSet<TaskTrack>();
+        }
+        
+        [Key]
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Data/UserTask.cs]: you don't need it, remove

File: DIMS/DIMS.DAL.Data/UserProfile.cs [hunk 6]
```diff
@@ -20,6 +21,7 @@ public UserProfile()
             this.UserTasks = new HashSet<UserTask>();
         }
     
+        [Key]
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Data/UserProfile.cs]: you don't need it, remove

File: DIMS/DIMS.DAL.Data/TaskTrack.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace HIMS.EF.DAL.Data
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel.DataAnnotations;
+
+    public partial class TaskTrack
+    {
+        [Key]
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Data/TaskTrack.cs]: you don't need it, remove

File: DIMS/DIMS.DAL.Identity/Models/ApplicationUser.cs [hunk 8]
```diff
@@ -9,5 +9,8 @@ namespace HIMS.EF.DAL.Identity.Models
 {
     public class ApplicationUser : IdentityUser
     {
+        public int Id { get; set; }
+        public string Email { get; set; }
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Identity/Models/ApplicationUser.cs]: you don't need it, remove; it already has all these fields

File: DIMS/DIMS.Tests/StubSampleRepository.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+                {
+                    sample = s;
+                }
+            }
+            
+            if (sample != null)
+            {
+                samples.Remove(sample);
+            }
+
```
Comment (maximsan) [inline DIMS/DIMS.Tests/StubSampleRepository.cs]: Use just LINQ methods, simplify this code

File: DIMS/DIMS.Tests/StubSampleRepository.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        public IEnumerable<Sample> Find(Func<Sample, bool> predicate)
+        {
+            //return samples.Find(predicate);
+            throw new Exception();
+        }
+
+        public Sample Get(int id)
+        {
+            throw new Exception();
```
Comment (maximsan) [inline DIMS/DIMS.Tests/StubSampleRepository.cs]: Complete code

File: DIMS/DIMS.Tests/StubSampleRepository.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public IEnumerable<Sample> GetAll()
+        {
+            return samples;
+        }
+
+        public void Update(Sample item)
+        {
+            throw new NotImplementedException();
```
Comment (maximsan) [inline DIMS/DIMS.Tests/StubSampleRepository.cs]: Complete code. I guess you can test it also.

**Comment by maximsan** [comment]
Please, clean cod, complete unit tests, remove all auto-generated code from business layer. Thanks!

File: DIMS/DIMS.BL/DTO/TaskDTO.cs (hunk 1 above)
Comment (VladRossohin) [reply DIMS/DIMS.BL/DTO/TaskDTO.cs]: fixing

File: DIMS/DIMS.BL/Interfaces/IUserService.cs (hunk 2 above)
Comment (VladRossohin) [reply DIMS/DIMS.BL/Interfaces/IUserService.cs]: done

File: DIMS/DIMS.BL/Services/UserProfileService.cs (hunk 3 above)
Comment (VladRossohin) [reply DIMS/DIMS.BL/Services/UserProfileService.cs]: done, I also add it in DirectionService

File: DIMS/DIMS.DAL.Identity/Models/ApplicationUser.cs (hunk 8 above)
Comment (VladRossohin) [reply DIMS/DIMS.DAL.Identity/Models/ApplicationUser.cs]: done

File: DIMS/DIMS.DAL.Data/TaskTrack.cs (hunk 7 above)
Comment (VladRossohin) [reply DIMS/DIMS.DAL.Data/TaskTrack.cs]: i done it, but now i have exceptions in mapping, so i think it's better to keep it with [Key] attribute
image

**Pull request merged**